[New Entry] C: data types (#1003)
Codecademy/docs: content/c/concepts/data-types/data-types.md

---
Title: 'Data Types'
Description: 'Data types are units of value stored in memory through variables, each with different functionalities and size-ranges.'
Subjects:
  - 'Computer Science'
  - 'Code Foundations'
Tags:
  - 'Booleans'
  - 'Variable Types'
  - 'Variables'
CatalogContent:
  - 'learn-c'
  - 'paths/computer-science'
---

> In C, **data types** are units of value stored in memory through [variables](https://www.codecademy.com/resources/docs/c/variables). Different kinds of data types determine their size and functionality.

Reviewer: ```suggestion
In C, **data types** are units of value stored in [memory](https://www.codecademy.com/resources/docs/c/memory-management) through [variables](https://www.codecademy.com/resources/docs/c/variables). Different kinds of data types determine their size and functionality.
```

## Static Typing

> Like with C++ and Java, C is a statically-typed language where variable types are known at compile time instead of the [runtime](https://www.codecademy.com/resources/docs/general/runtime). In this case, types must be expressly indicated by the programmer.

Reviewer 2: Do we mean that types are explicitly indicated?

Data type classifications supported in C include the following:

- Primitive data types
- User-defined data types
- Derived data types

Types also differ in the amount of memory they occupy and the range of values they can hold. This will depend on the compiler of the computer. An integer, for example, might take up 32 bits on one computer, or perhaps it might be stored in 64 bits on another.

## Primitive types

> These data types are mostly numeric (integers, doubles, floating-point numbers, etc.) and can be used for arithmetic operations like subtraction and division. Values that use the `void` keyword (for no data) are also considered to be primitive data types.

Reviewer: ```suggestion
These data types are mostly numeric (integers, doubles, floating-point numbers, etc.) and can be used for arithmetic operations like subtraction and division. Values that use the `void` keyword (for no data) are also regarded as primitive data types.
```

## Derived data types

These data types derive from primitive or fundamental data types. In C, these can be of three types:

- [Arrays](https://www.codecademy.com/resources/docs/c/arrays)
- [Functions](https://www.codecademy.com/resources/docs/c/functions)
- Pointers

## User-defined types

These are a collection of data with values that have fixed meanings that extend the primitive types already available. Users define them based on requirements. C consists of the following user-defined types:

- Enum
- [Structures](https://www.codecademy.com/resources/docs/c/structures)
- Typedef
- Union

## Memory Size

| Type   | Size    | Value Range                     |
| ------ | ------- | ------------------------------- |

> | Char   | 1 byte  | -128 to 127 or 0 to 255         |
> | Int    | 4 bytes | -2,147,483,648 to 2,147,483,647 |
> | Float  | 4 bytes | 1.2E-38 to 3.4E+38              |
> | Double | 8 bytes | 1.7E-308 to 1.7E+308            |

Reviewer: ```suggestion
| Char   | 1 byte  | -128 - 127 or 0 - 255           |
| Int    | 4 bytes | -2,147,483,648 - 2,147,483,647  |
| Float  | 4 bytes | 1.2E-38 - 3.4E+38               |
| Double | 8 bytes | 1.7E-308 - 1.7E+308             |
```

> **Note:** Hexadecimal integer literals begin always with the `0` followed by either an `x` or `X`, followed by different combinations of digits (`0` - `9`) and letters (`a` - `f` or `A` - `F`).

```c
int = 0;
int = -5;
int = 20;
int rgbColor = 0XFFEF0D; // Hexadecimal
```

Spaces aren't allowed between digits and the values larger than 999 don't use commas.

```c
int hundreds = 12,000; // This is wrong
int hundreds2 = 12000 // This is correct
```

### Float and Double

Floating-point values are non-whole numbers that contain decimal places. They can have up to 6 decimal digits of precision and can be expressd in scientific notation.

> **Note:** Scientific notation is a form of expressing numbers that are too large or too small to be conveniently written in decimal form (e.g., `6.022E23`).

Doubles are like floats but more precise with up to 15 decimal digits. The following is an example of `float` and `double` type values:

```c
float salary = 523.45;
double bankTransaction = 245.2518;
float counterFactor = 15.87E4;
```

### Char

> Chars use single quotations marks `'` to present single characters such as letters, [numbers](https://www.codecademy.com/resources/docs/general/number), [whitespace](https://www.codecademy.com/resources/docs/general/whitespace), special characters, and symbols. Each char only occupies one byte in memory. In C, a sequence of chars is called [strings](https://www.codecademy.com/resources/docs/c/strings).

Reviewer: ```suggestion
Chars use single quotations marks `'` to represent single characters such as letters, [numbers](https://www.codecademy.com/resources/docs/general/number), [whitespace](https://www.codecademy.com/resources/docs/general/whitespace), special characters, and symbols. Each char only occupies one byte in memory. In C, a sequence of chars is called [strings](https://www.codecademy.com/resources/docs/c/strings).
```

> Chars use single quotations marks `'` to represent single characters such as letters, [numbers](https://www.codecademy.com/resources/docs/general/number), [whitespace](https://www.codecademy.com/resources/docs/general/whitespace), special characters, and symbols. Each char only occupies one byte in memory. In C, a sequence of chars is called [strings](https://www.codecademy.com/resources/docs/c/strings).

Reviewer 3: ```suggestion
Chars use single quotations marks `'` to represent single characters such as letters, [numbers](https://www.codecademy.com/resources/docs/general/number), [whitespace](https://www.codecademy.com/resources/docs/general/whitespace), special characters, and symbols. Each char only occupies one byte in memory. In C, sequences of chars are called [strings](https://www.codecademy.com/resources/docs/c/strings).
```

---
Title: 'Data Types'

> Description: 'Java’s most basic data types are known as primitive data types and are in the system by default. The available types are as follows: - int - char - boolean - byte - long'

Reviewer 3: ```suggestion
Description: 'Java’s most basic data types are known as primitive data types and are in the system by default. The available types are as follows: - int - char - boolean - byte - long'
```
```suggestion
Description: 'The most basic data types in C are known as primitive data types and are in the system by default.'
```

Reviewer 3: Removed reference to Java. Also, descriptions are best as one sentence.

Subjects:
  - 'Computer Science'
Tags:
  - 'Data Types'
  - 'Integers'
  - 'Booleans'
CatalogContent:
  - 'learn-c'
  - 'paths/computer-science'
---

## Basic Data Types

C supports many different types of variables and each type of variable is used for storing some type of data.

Some examples of basic data types in C are:

- `int` - numbers (non-decimal)
- `float` - decimal numbers
- `double` - decimal numbers (stores more data than float)
- `char` - characters

The difference between the types is in the amount of memory they occupy and the range of values they can hold. This will depend on the computer running since it's machine dependent. An integer might take up 32bits on one computer, or perhaps it might be stored in 64bits.

### Data types examples

```c
int age = 25;
float = 125.8;
double = 3.2325467;
char = 'T';
```

### Int

> It contains only integral values. A minus sign preceding the data type indicates that the value is negative.

Reviewer 3: ```suggestion
Ints contain only integral values. A minus sign preceding the data type indicates that the value is negative.
```

> Can be expressed in hexadecimal notation (base 16).

Reviewer 3: ```suggestion
They can be expressed in hexadecimal notation (base 16).
```

Reviewer 3: I'd add an explanation of the syntax for hexadecimal values.

```c
int = 0;
int = -5;
int = 20;
int rgbColor = 0XFFEF0D;
```

> Spaces aren't allowed between digits and the values larger than 999 don't need a comma.

Reviewer 3: ```suggestion
Spaces aren't allowed between digits and the values larger than 999 don't use commas.
```

```c
int hundreds = 12,000; // This is wrong
int hundreds2 = 12000 // Correct!
```

### Float and Double

> Floats are used to store floating-point numbers, values that contain decimal plates. They can be expressed in scientific notation and has 6 decimal digits of precision.

Reviewer 3: ```suggestion
Floats are used to store floating-point numbers, values that contain decimal places. They can be expressed in scientific notation and have 6 decimal digits of precision.
```

Reviewer 3: I'd add a sentence explaining the syntax of scientific notation.

> Doubles are the same but they are needed when more precision is required since it has 15 decimal digits of precision.

Reviewer 3: ```suggestion
Doubles are the same but they are used when more precision is required. They have 15 decimal digits of precision.
```

```c
float salary = 523.45;
double bankTransaction = 245.2518;
float counterFactor = 15.87e4;
```

> 

Reviewer 3: We're missing a section on the `char` datatype.

## Memory Size

| Type   | Size    | Value Range                     | Format Specifier |
| ------ | ------- | ------------------------------- | ---------------- |
| Char   | 1 byte  | -128 to 127 or 0 to 255         | %c               |
| Int    | 4 bytes | -2,147,483,648 to 2,147,483,647 | %d               |
| Float  | 4 bytes | 1.2E-38 to 3.4E+38              | %f               |

> | Double | 8 bytes | 1.7E-308 to 1.7E+308            | %lf              |

Reviewer 3: There's no mention of what the "Format Specifier" means in this entry. That column should probably be omitted.

---
Title: 'Data Types'

> Description: 'C’s most basic data types are known as primitive data types and are in the system by default.'

Reviewer: ```suggestion
Description: 'Data types are units of value stored in memory through variables, each with different functionalities and size-ranges.'
```

Subjects:
  - 'Computer Science'
Tags:
  - 'Data Types'
  - 'Integers'

>   - 'Booleans'

Author: I should always write the tags in order, right? For a matter of consistency?

>   - 'Booleans'

Reviewer: Nope! Order does not matter 😄

CatalogContent:
  - 'learn-c'
  - 'paths/computer-science'
---

## Basic Data Types

C supports many different types of variables and each type of variable is used for storing some type of data.

Some examples of basic data types in C are:

- `int` - numbers (non-decimal)
- `float` - decimal numbers
- `double` - decimal numbers (stores more data than float)
- `char` - characters

The difference between the types is in the amount of memory they occupy and the range of values they can hold. This will depend on the computer running since it's machine dependent. An integer might take up 32bits on one computer, or perhaps it might be stored in 64bits.

### Data types examples

```c
int age = 25;
float = 125.8;
double = 3.2325467;
char = 'T';
```

### Int

> Ints contains only integral values. A minus sign preceding the data type indicates that the value is negative.
> They can be expressed in hexadecimal notation (base 16).\*

Reviewer: ```suggestion
Ints contains only integral values. A minus sign preceding the data type indicates that the value is negative. They can be expressed in hexadecimal notation (base 16).
```

> \* Hexadecimal integer literals begin always with the 0 followed by either an `x` or `X`, followed by different combinations of digits going from `0` through `9`, and the letters `a` through `f` or `A` through `F`.

Reviewer: ```suggestion
> **Note:** Hexadecimal integer literals begin always with the `0` followed by either an `x` or `X`, followed by different combinations of digits (`0` - `9`) and letters (`a` - `f` or `A` - `F`).
```

```c
int = 0;
int = -5;
int = 20;
int rgbColor = 0XFFEF0D; // Hexadecimal
```

Spaces aren't allowed between digits and the values larger than 999 don't use commas.

```c
int hundreds = 12,000; // This is wrong

> int hundreds2 = 12000 // Correct!

Reviewer: ```suggestion
int hundreds2 = 12000; // This is correct
```

```

### Float and Double

> Floats are used to store floating-point numbers, values that contain decimal places. They can be expressed in scientific notation\* and has 6 decimal digits of precision.

Reviewer: ```suggestion
Floating-point values are non-whole numbers that contain decimal places. They can have up to 6 decimal digits of precision and can be expressed in scientific notation.
```

> \*Scientific notation is a form of expressing numbers that are too large or too small to be conveniently written in decimal form. E.g. `6.022E23`

Reviewer: ```suggestion
> **Note:** Scientific notation is a form of expressing numbers that are too large or too small to be conveniently written in decimal form ( e.g., `6.022E23`).
```

> Doubles are the same but they are used when more precision is required. They have 15 decimal digits of precision.

Reviewer: ```suggestion
Doubles are like floats but more precise with up to 15 decimal digits. The following is an example of `float` and `double` type values:
```

```c
float salary = 523.45;
double bankTransaction = 245.2518;

> float counterFactor = 15.87e4;

Reviewer: ```suggestion
float counterFactor = 15.87E4;
```

Reviewer: Not sure if capitalization matters but suggesting uppercase "E" for consistency.

```

### Char

> Chars allows its variables to store only a single character a letter, number, space, punctuation mark or symbol. Single quotes are used to represent them. They only occupy one byte in memory.
> In C, a sequence of chars are called [strings](https://www.codecademy.com/resources/docs/c/strings).

Reviewer: ```suggestion
Chars use single quotation marks `'` to present single characters such as letters, [numbers](https://www.codecademy.com/resources/docs/general/number), [whitespace](https://www.codecademy.com/resources/docs/general/whitespace), special characters, and symbols. Each char only occupies one byte in memory. In C, a sequence of chars are called [strings](https://www.codecademy.com/resources/docs/c/strings).
```

```c
char a = 'a';
```

## Memory Size

| Type   | Size    | Value Range                     |
| ------ | ------- | ------------------------------- |
| Char   | 1 byte  | -128 to 127 or 0 to 255         |
| Int    | 4 bytes | -2,147,483,648 to 2,147,483,647 |
| Float  | 4 bytes | 1.2E-38 to 3.4E+38              |
| Double | 8 bytes | 1.7E-308 to 1.7E+308            |

> 

Reviewer: I would remove this section altogether and, instead, place this table right above the "### Int" header.

## Static Typing

> C is a statically-typed language (such as C++ and Java) where variable types are known at compile time instead of the runtime. In this case, types must be expressly indicated by the programmer.

Reviewer: This section should be moved somewhere near the top of this file (in lines # 15-18).
